src/VisualStudio/LiveShare/Impl/Client/RemoteLanguageServiceWorkspace.cs

```diff
@@ -97,97 +113,123 @@ void IRunningDocumentTableEventListener.OnRenameDocument(string newMoniker, stri
+            }
         }
 
-        public void Init()
+        /// <summary>
+        /// Retrieves the base folder paths for files on the client that have been retrieved from the remote host.
+        /// Triggers a refresh of all open files so we make sure they are in the correct workspace.
+        /// </summary>
+        private async Task UpdatePathsToRemoteFilesAsync()
         {
-            StartSolutionCrawler();
-        }
+            if (_session == null)
+            {
+                return;
+            }
 
-        /// <inheritdoc />
-        public override bool CanOpenDocuments => true;
+            var roots = await _session.ListRootsAsync(CancellationToken.None).ConfigureAwait(false);
+            if (roots.Length > 0)
+            {
+                var localPathsOfRemoteRoots = roots.Select(root => _session.ConvertSharedUriToLocalPath(root)).ToImmutableArray();
 
-        /// <inheritdoc />
-        public void OnManagedProjectAdded(ProjectInfo projectInfo)
-        {
-            //Notify AnyCode Roslyn workspace a project is being added.
-            OnProjectAdded(projectInfo);
-        }
+                var remoteRootPaths = new HashSet<string>();
+                var externalPaths = new HashSet<string>();
 
-        /// <inheritdoc />
-        public void OnManagedProjectReloaded(ProjectInfo projectInfo)
-        {
-            //Notify AnyCode Roslyn workspace a project is being reloaded.
-            OnProjectReloaded(projectInfo);
+                foreach (var localRoot in localPathsOfRemoteRoots)
+                {
+                    var remoteRootPath = localRoot.Substring(0, localRoot.Length - 1);
+                    var lastSlash = remoteRootPath.LastIndexOf('\\');
+                    var externalPath = remoteRootPath.Substring(0, lastSlash + 1);
+                    externalPath += "~external";
+
+                    remoteRootPaths.Add(remoteRootPath);
+                    externalPaths.Add(externalPath);
+                }
+
+                if (!_remoteWorkspaceRootPaths.Equals(remoteRootPaths) || !_registeredExternalPaths.Equals(externalPaths))
+                {
+                    _remoteWorkspaceRootPaths = remoteRootPaths.ToImmutableHashSet();
+                    _registeredExternalPaths = externalPaths.ToImmutableHashSet();
+                    await RefreshAllFiles().ConfigureAwait(false);
+                }
+            }
         }
 
-        /// <inheritdoc />
-        public void OnManagedProjectReferenceAdded(ProjectId projectId, ProjectReference reference)
+        public void EndSession()
         {
-            //Notify AnyCode Roslyn workspace a project reference is being reloaded.
-            OnProjectReferenceAdded(projectId, reference);
+            IsRemoteSession = false;
         }
 
-        /// <inheritdoc />
-        public void UpdateProjectReferences(ProjectId projectId, IEnumerable<ProjectReference> references)
+        public void Init()
```

Comment:
(Random nitpick not related to your change: it's odd Init is called here from RemoteLanguageServiceWorkspaceHost when we could have just written this in SetSession)